apps/inheritance_app/inheritance_encrypt_data.c

```diff
@@ -0,0 +1,497 @@
+  pb_size_t index = 0;
+
+  for (index = 0; index < context->data_count - 1; index++) {
+    inheritance_fill_tlv(context->packet,
+                         &context->packet_size,
+                         0x00,
+                         context->data[index].encrypted_data_size,
+                         context->data[index].encrypted_data);
+  }
+
+  // The last encrypted message is the PIN
+  inheritance_fill_tlv(context->packet,
+                       &context->packet_size,
+                       0x50,
+                       context->data[index].encrypted_data_size,
+                       context->data[index].encrypted_data);
+
+  return true;
+}
+
+static bool encrypt_packet(void) {
+  if (SESSION_ENCRYPT_PACKET_SUCCESS !=
+      session_aes_encrypt(context->packet, &context->packet_size)) {
+    return false;
+  }
+
+  return true;
+}
+
+static bool encrypt_data(void) {
+  bool status = true;
+
+  do {
+    if (!inheritance_verify_pin()) {
+      // TODO: Throw user rejceted
+      status = false;
+      break;
+    }
+    if (!serialize_message_data()) {
+      // TODO: Throw serialization failed
+      status = false;
+      break;
+    }
+
+    if (!encrypt_message_data()) {
+      // TODO: Throw encryption failed
+      status = false;
+      break;
+    }
+
+    if (!serialize_packet()) {
+      // TODO: Throw packet serialization error
+      status = false;
+      break;
+    }
+
+    if (!encrypt_packet()) {
+      // TODO: Throw packet encryption error
+      status = false;
+      break;
+    }
+  } while (0);
+
+  return status;
+}
+
+static bool send_encrypted_data(inheritance_query_t *query) {
+  inheritance_result_t result =
+      init_inheritance_result(INHERITANCE_RESULT_ENCRYPT_TAG);
+  result.encrypt.which_response =
+      INHERITANCE_ENCRYPT_DATA_WITH_PIN_RESPONSE_RESULT_TAG;
+  if (!inheritance_get_query(query, INHERITANCE_QUERY_ENCRYPT_TAG) ||
+      !check_which_request(
+          query, INHERITANCE_ENCRYPT_DATA_WITH_PIN_REQUEST_INITIATE_TAG)) {
+    return false;
+  }
+
+  result.encrypt.result.encrypted_data.size = context->packet_size;
+  memcpy(&result.encrypt.result.encrypted_data.bytes,
+         context->packet,
+         context->packet_size);
+
+  inheritance_send_result(&result);
+  return true;
+}
+
+/*****************************************************************************
+ * GLOBAL FUNCTIONS
+ *****************************************************************************/
+
+void inheritance_encrypt_data(inheritance_query_t *query) {
+  context = (inheritance_encryption_context_t *)malloc(
+      sizeof(inheritance_encryption_context_t));
+  ASSERT(context != NULL);
+  memzero(context, sizeof(inheritance_encryption_context_t));
+
+  if (inheritance_encryption_handle_inititate_query(query) &&
+      inheritance_encryption_get_user_verification() && encrypt_data() &&
+      send_encrypted_data(query)) {
+    delay_scr_init(ui_text_check_cysync, DELAY_TIME);
+  }
+
+  free(context);
```

Comment:
preferably memzero `context` before free as it has private data, or could clear them right after they are no longer needed